### PR TITLE
Implement global search

### PR DESCRIPTION
**Pull request by koying**
**Changes**
This adds leanback global search support

image

image

### CODE REVIEW

**Comment by nielsvanvelzen** [comment]
I was planning to add this over a year ago but the app wasn't ready to process api calls in the background. Your current implementation looks good but there's a few things that need to be done before I'm willing to merge it.

Right now we have the LeanbackChannelWorker which updates the "next up" and other channels in the background. It uses the currently signed in user. But it should use the system user as set in the authentication preferences instead. Right now this is not possible because there is no authenticated api client ready to use. My plan was to work on that part after #880 is merged.

When that is ready this PR could be updated to use the new api client instance (which will use the new SDK). I'm blocking this PR for now until that's ready. Sorry for the inconvenience. 



Also, it looks like some files are not indented properly (some lines use spaces for example) so please reformat the code (can be easily done with Android Studio).

**Comment by koying** [comment]
> But it should use the system user as set in the authentication preferences instead

Not really sure what the "system" user is used for. I assume it's for authenticated, non-user specific, data?
But you want the channel data and search to show the data available to the signed-in user anyway, don't you?

You don't want the search to return, e.g., a movie, that the signed-in user can't actually watch.

**Comment by nielsvanvelzen** [comment]
Most of the time the system user is the currently authenticated user, but it's possible to always use a specific user or no user (to disable integrations). This is done for cases where you want to have a "safe" account to provide media to the system.

> You don't want the search to return, e.g., a movie, that the signed-in user can't actually watch.

This can always happen if you switch to a different user when opening the app, but the system user is supposed to help with this issue by allowing the user to change which account to use.

**Comment by nielsvanvelzen** [comment]
With #912 now merged you can rebase this PR on top of master and use the systemApiClient, look at the LeanbackChannelWorker to see how it works.

**Comment by koying** [comment]
@nielsvanvelzen I'm at loss at what has to be done, I'm afraid.
By what am I supposed to replace 

```
suspend fun ApiClient.getItem(id: String): BaseItemDto? = suspendCoroutine { continuation ->
```

by ?

https://github.com/jellyfin/jellyfin-androidtv/blob/master/app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt#L49

https://github.com/jellyfin/jellyfin-androidtv/pull/900/files#diff-0af690fd0eb4e4dec46d25d886975c10b8398a48538be6789eded50fa092dc55R87

**Comment by nielsvanvelzen** [comment]
In short: everything using the `org.jellyfin.apiclient` package should use the `org.jellyfin.sdk` package. We've rewritten the entire apiclient (which was basically C# code transformed into Java, it was a mess) into 100% Kotlin with auto-generated code using OpenAPI specs. The app should eventually move entirely to this new SDK and everything running in the background / as system user is now required to use it.

The `getItem` call in the ApiClient should use the `getItem` function in the `UserLibraryApi` (https://api.jellyfin.org/#operation/GetItem)

**Comment by koying** [comment]
Thanks for the explanation 😊

**Comment by koying** [comment]
@nielsvanvelzen Rebased and refactored as requested

File: app/src/main/res/xml/searchable.xml
Comment (nielsvanvelzen) [inline app/src/main/res/xml/searchable.xml:1]: Please re-format the file in Android Studio

File: app/src/main/res/xml/searchable.xml
Comment (nielsvanvelzen) [inline app/src/main/res/xml/searchable.xml:4]: I think we want a different string for this one as the current one says "Search text (select for keyboard)". The Android docs  recommend something like "Search Jellyfin" or "Search your media" etc.

https://developer.android.com/guide/topics/search/searchable-config

File: app/src/main/res/xml/searchable.xml
Comment (nielsvanvelzen) [inline app/src/main/res/xml/searchable.xml:7]: The path shouldn't be necessary according to the docs

File: app/src/main/res/xml/searchable.xml
Comment (nielsvanvelzen) [inline app/src/main/res/xml/searchable.xml:13]: I think this should be a `@string` resource

File: app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt:21]:
```suggestion
			Timber.d("LAUNCH: %s", data)
```

File: app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt:36]:
```suggestion
		if (handleIntent(intent)) return
```

File: app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt:55]:
```suggestion
		handleIntent(intent);
```

File: app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/search/SearchActivity.kt:17]: I'd prefer having this function after onNewIntent (where it's used)

File: app/src/main/AndroidManifest.xml
Comment (nielsvanvelzen) [inline app/src/main/AndroidManifest.xml:138]: It might be better to open the MainActivity with the search intent instead of the SearchActivity

Comment (nielsvanvelzen) [reply app/src/main/AndroidManifest.xml:138]: this would also move the search intent filter

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:1]: Move to integration package

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:1]: and reformat with Android Studio

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:42]: Move to companion object

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:49]:
```suggestion
	private val sURIMatcher = UriMatcher(UriMatcher.NO_MATCH).apply {
		addURI(AUTHORITY, SUGGEST_PATH + "/" + SearchManager.SUGGEST_URI_PATH_QUERY, SEARCH_SUGGEST)
		addURI(AUTHORITY, SUGGEST_PATH + "/" + SearchManager.SUGGEST_URI_PATH_QUERY + "/*", SEARCH_SUGGEST)
	}
```

Can also be moved to companion object

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:53]:
```suggestion
	override fun onCreate() = true
```

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:53]: Actually, instead of always returning `true` it should return the usability state of the api client: `apiClient.isUsable`

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:56]: Because of the amount of parameters it's more readable to put each one on it's own line:
```suggestion
	override fun query(
		uri: Uri,
		projection: Array<String>?,
		selection: String?,
		selectionArgs: Array<String>?,
		sortOrder: String?,
	): Cursor? {
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:64]:
```suggestion
				val limit = uri.getQueryParameter("limit")?.toIntOrNull() ?: 10
```

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:64]: Might want to move "limit" to a constant like "QUERY_PARAM_LIMIT"

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:65]:
```suggestion
				return uri.lastPathSegment?.let { getSuggestions(it.lowercase(), limit) }
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:90]:
```suggestion
			fields = listOf(
				ItemFields.DATE_CREATED,
				ItemFields.STUDIOS,
				ItemFields.GENRES,
				ItemFields.TAGLINES,
				ItemFields.PROVIDER_IDS,
				ItemFields.MEDIA_STREAMS
			)
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:83]:
```suggestion
				sortBy = listOf("DateCreated"),
```

`ItemSortBy` is from the old apiclient, the SDK doesn't have an enum or constant for it right now.

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:83]: You could also use `ItemFields.DATE_CREATED.serialName` or `ItemFields.DATE_CREATED.toString()`

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:74]:
```suggestion
	private suspend fun searchItems(query: String, limit: Int): BaseItemDtoQueryResult? {
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:78]:
```
val items by itemsApi.getItems(...))
return items
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:95]:
```suggestion
		val searchResult = searchItems(query, limit)
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:98]:
```suggestion
		if (searchResult != null) Timber.d("Search returned %d items", searchResult.totalRecordCount)
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:114]:
```suggestion

		val cursor= MatrixCursor(menuCols)
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:100]:
```suggestion
		val columns = arrayOf(
```

Might be worth making that one a constant too?

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:123]: Use `.toUri()` instead

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:120]: WEBP is preferred

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:134]:
```suggestion
			mc.addRow(arrayOf(
```

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:134]: Can you add comments to all parameters like `// SUGGEST_COLUMN_TEXT_1` etc. so it's easier to read which columns the values are for?

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:157]:
```suggestion
	override fun getType(uri: Uri) = null
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:161]:
```suggestion
	override fun insert(uri: Uri, values: ContentValues?) = null
```

File: app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/content/MediaContentProvider.kt:169]: Same as the 2 above, just make it no-op instead of throwing an exception

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

**Comment by nielsvanvelzen** [comment]
Hey @koying, now that 0.12 is released we're unfreezing the master branch again. Are you still interested in finishing this pull request?

**Comment by koying** [comment]
I'll try to get back to it, but if you feel like taking the hand (or just redo it your way), don't hesitate.
Kotlin is a tad chinese to me, I must say ,)

**Comment by nielsvanvelzen** [comment]
No problem, I might pick it up in a week or so when the other PR's are merged. Thanks for the initial code anyway!

**Comment by nielsvanvelzen** [comment]
(I've only rebased this PR for now, will fix the review feedback later)

**Comment by nielsvanvelzen** [comment]
#1233 fixes the startup issue. I looked into the image loading issue I experienced and found out it was because my server is served over HTTP locally. When the URL is https the images are shown as expected.

Also, I just realized I amended my changes to your commit @koying. My intention was to add my changes in a separate commit so my apologies for that.

_Note that without the changes from #1233 this PR will probably crash instantly_

**Comment by phirestalker** [comment]
How much codebase is shared between the TV apps? Will this enable it in the Tizen app, or will it need to be coded there too?

**Comment by nielsvanvelzen** [comment]
The Tizen app is a completely different project

**Review by nielsvanvelzen — APPROVED** [review]

**Review by thornbill — COMMENTED** [review]
This is crashing on startup for me:
```
11-16 23:52:52.444 23117 23117 E ACRA    : ACRA caught a RuntimeException for org.jellyfin.androidtv.debug
11-16 23:52:52.444 23117 23117 E ACRA    : java.lang.RuntimeException: Unable to get provider org.jellyfin.androidtv.integration.MediaContentProvider: java.lang.IllegalStateException: KoinApplication has not been started
11-16 23:52:52.444 23117 23117 E ACRA    :      at android.app.ActivityThread.installProvider(ActivityThread.java:5931)
```

**Comment by mueslimak3r** [comment]
> This is crashing on startup for me:
> 
> ```
> 11-16 23:52:52.444 23117 23117 E ACRA    : ACRA caught a RuntimeException for org.jellyfin.androidtv.debug
> 11-16 23:52:52.444 23117 23117 E ACRA    : java.lang.RuntimeException: Unable to get provider org.jellyfin.androidtv.integration.MediaContentProvider: java.lang.IllegalStateException: KoinApplication has not been started
> 11-16 23:52:52.444 23117 23117 E ACRA    :      at android.app.ActivityThread.installProvider(ActivityThread.java:5931)
> ```

> #1233 fixes the startup issue. I looked into the image loading issue I experienced and found out it was because my server is served over HTTP locally. When the URL is https the images are shown as expected.
> 
> Also, I just realized I amended my changes to your commit @koying. My intention was to add my changes in a separate commit so my apologies for that.
> 
> _Note that without the changes from #1233 this PR will probably crash instantly_

Are you testing it with the changes in PR 1233?

**Comment by thornbill** [comment]
Yeah I rebased my local branch on master just to make sure it was current.

**Comment by nielsvanvelzen** [comment]
I was able to confirm the issue with an Android TV 8 based emulator. The fix was easy: adding `initOrder` attributes to the contentproviders for both androidx.startup and the MediaContentProvider.
Please try again @thornbill

**Review by thornbill — APPROVED** [review]
:rocket: confirmed the crash is fixed!

**Pull request merged**
